kythe/docs/rfc/2909.md

```diff
@@ -0,0 +1,145 @@
+---
+title: Kythe Columnar Serving Format
+date: 2018/08/08
+link: https://github.com/google/kythe/pulls/2909
+---
+
+# Background
+
+The current Kythe serving data format is just a simple key-value store from string keys (usually [Kythe tickets](https://kythe.io/docs/kythe-uri-spec.html)) to ProtocolBuffer message.  This format requires the Kythe PostProcessor to perform a complex reduction over a potentially massive number of decorations/references/edges/etc. to produce a single lookup value per node.  A set of CrossReferences can be so large that it already requires manual paging, another complex transformation in the PostProcessor.  These operations tend to create hot shards and stragglers, leading to long post-processing times and potentially failures due to memory exhaustion.  Likewise, the server is forced to slowly decode large ProtocolBuffer messages even when only a subset of the data is required and the large value sizes can cause issues with some backend stores.
```

Comment:
…to ProtocolBuffer message values.

Reply:
Maybe consider wrapping the text to some screen-ish width? That would make it easier to comment on in the diff view.

Reply:
✓